crates/interpreter/src/instructions/memory.rs

```diff
@@ -35,3 +42,31 @@ pub fn msize(interpreter: &mut Interpreter, _host: &mut dyn Host) {
     gas!(interpreter, gas::BASE);
     push!(interpreter, U256::from(interpreter.memory.effective_len()));
 }
+// From EIP-5656 MCOPY
+pub fn mcopy<SPEC: Spec>(interpreter: &mut Interpreter, _host: &mut dyn Host) {
+    // check that Cancun spec is enabled
+    check!(interpreter, SPEC::enabled(CANCUN));
+    // get src and dest and length from stack
+    pop!(interpreter, dest, src, len);
+
+    // into usize or fail
+    let len = as_usize_or_fail!(interpreter, len, InstructionResult::InvalidOperandOOG);
+    let dest = as_usize_or_fail!(interpreter, dest, InstructionResult::InvalidOperandOOG);
+    let src = as_usize_or_fail!(interpreter, src, InstructionResult::InvalidOperandOOG);
+    // deduce gas
+    gas_or_fail!(interpreter, gas::verylowcopy_cost(len as u64));
+
+    // check if len is 0
+    if len == 0 {
+        return;
+    }
```

Comment:
There is a bug here. in order of checks that are done.

If `len == 0` we should just take BASE gas and return, but if `dest` or `src` are over `usize` we would have a consensus error and report `InvalidOperatorOOG`.

The order should be:

```suggestion
    let len = as_usize_or_fail!(interpreter, len, InstructionResult::InvalidOperandOOG);
    // deduce gas
    gas_or_fail!(interpreter, gas::verylowcopy_cost(len as u64));
    if len == 0 {
        return;
    }
    
    let dest = as_usize_or_fail!(interpreter, dest, InstructionResult::InvalidOperandOOG);
    let src = as_usize_or_fail!(interpreter, src, InstructionResult::InvalidOperandOOG);
```